pycardano/serialization.py

```
@@ -240,6 +265,8 @@ def to_primitive(self) -> Primitive:
         def _dfs(value, freeze=False):
             if isinstance(value, CBORSerializable):
                 return _dfs(value.to_primitive(), freeze)
+            elif isinstance(value, bytes):
+                return ByteString(value)
```

IMO, it seems incorrect to replace every bytes with ByteString. Instead, we just offer ByteString for users to use in PlutusData or Metadata. 
For some internal implementation that generates bytes as intermediate values, e.g. [script_data_hash](https://github.com/Python-Cardano/pycardano/blob/eaadc385e10f94355d1e632e3c75c9c3f5e3e46c/pycardano/utils.py#L212), we don't want to change its type arbitrarily.

---

Should be implementable by a simple parameter?

---

That was one of my original points and how I originally had it implemented. 

Can someone please make a definitive final decision so I can fix and be done? I've implemented and reimplemented this multiple times.

---

@nielstron Could you elaborate how adding a parameter will work?

---

It might not work as straightforward as I imagined it. @theeldermillenial was right, maybe we should just roll with the initial design. I appreciate the excourse though because now we know precisely which bytes to encode this way 😅 sorry for the divergence.

---

You just earned my respect with that. I appreciate it.

So here's what I propose. We create a variation on the original idea of a dummy class for `PlutusData`. Probably use that for metadata as well?

I would prefer that we try to get it right this time so we can finish it off. If someone wants to tackle it, that's okay. I'm happy to work on it, but I want to hear a consensus before starting to work on it this time around.

---

My preferred approach is to offer users `ByteString` class to use, which the encoder can automatically break it down to byte array. If a bytes object is found longer than 64 instead, pycardano should raise an exception and recommend users to use `ByteString`.

---

@cffls But should this error only be thrown for `PolusData` and Metadata? Or should we apply it globally?

My two cents is "only implement exactly what is defined". The bytes length restriction appears to be limited to "metadata", so maybe we only apply it to metadata.

---

We already have this check for metadata: https://github.com/Python-Cardano/pycardano/blob/main/pycardano/metadata.py#L40-L49

I thought this should be also enforced in PlutusData, which was the reason why this PR was raised. If not, I am fine with only providing ByteString as an option for user in this PR.

---

My apologies. I misspoke. When I said Metadata, I also meant PlutusData.

Also, I now see exactly what you're saying, and I think you're solution makes the most sense. You are saying we should inject a check for long byte strings in PlutusData and throw an error similar to what is seen in Metadata. Part of that error message should indicate the user can use the new ByteString class to allow longer bytes.

I think this is the most transparent approach, and it keeps in line with what I see to be pycardano's philosophy of being very unbiased.

If this is what you mean, I'll make the changes and we can be done. I will revert any hashes I altered, since this should really only affect the test I created. If there's any other unit tests you would like to see, I'm happy to add them.

---

Yes, this is exactly what I meant. Please go ahead with this approach. Thank you for confirming. ☺️